### PR TITLE
add methods getTiles and setTiles to mimick maplibre-gl-js interface

### PR DESCRIPTION
**Pull request by xabbu42**
This adds the methods setTiles and getTiles to VectorSource to mimick the maplibre-gl-js interface
introduced in https://github.com/mapbox/mapbox-gl-js/pull/8048. This is useful to change parametrized data sources without having to load a complete new style.

Todo:
- [ ] To completely port the maplibre-gl-js interface, also setUrl should be added.
- [ ] Is the additional call to observer->onSourceLoaded correct/necessary?
- [ ] There is a race and the map sometimes still requests old tiles even after setTile is called.

Any input especially about the race is appreciated.

### CODE REVIEW

**Comment by xabbu42** [comment]
We use this patch in production for over a year now. I just finally got around to update to latest maplibre-native opengl-2 branch and it still works as expected. Moreover I can no longer reproduce the race and I doubt it was a problem with this patch anyway.

I no longer thinks we should block this on an implementation of setUrl. That is a separate issue and a bit more complex, so I'd rather just get this in for now.

From my reading of the code the call to `observer->onSourceLoaded` is not necessary for Core funktionality, as `onSourceLoaded` and `onSourceChanged` ist handled exactly the same as `onSourceLoaded` in https://github.com/maplibre/maplibre-native/blob/main/src/mbgl/style/style_impl.cpp#L314. I'm happy to leave it in or take it out whatever the semantics of an external Observer are supposed to be.

So this ready for reviewing from my point of view.

**Comment by wipfli** [comment]
Nice @xabbu42! Please remove the draft status in that case

**Comment by xabbu42** [comment]
I'll have to look in the failed checks tomorrow.

**Review by louwers — COMMENTED** [review]
Thanks for your contribution.

Could you add:

- A test
- A doxygen docstring (since these methods are part of the public API)

**Comment by xabbu42** [comment]
Sorry for the long silence. I'll be away for the next 3 weeks but I plan to get back to this after that.

**Comment by xabbu42** [comment]
The checks run through before the merge and it looks like the failures now have nothing to do with my changes. @louwers I hope I addressed your requests adequately (I'm not familiar with doxygen or GoogleTest). So this should be ready to merge now.

**Comment by louwers** [comment]
@xabbu42 Seems like it runs through cleanly now. Let me have another look.

File: include/mbgl/style/sources/vector_source.hpp
Comment (louwers) [inline include/mbgl/style/sources/vector_source.hpp:31]: Should be marked `const`.

**Review by louwers — CHANGES_REQUESTED** [review]
Two more nits, the getter can be marked `const`. And you can remove the type from the Doxygen comments. Then it looks good to merge!

File: include/mbgl/style/sources/vector_source.hpp
Comment (xabbu42) [reply include/mbgl/style/sources/vector_source.hpp:31]: I removed the type from the docstring. The return type already is marked `const` or am I missing something?

Comment (louwers) [reply include/mbgl/style/sources/vector_source.hpp:31]: Right, but the method itself can also be marked const right? It does not modify the object.

Comment (xabbu42) [reply include/mbgl/style/sources/vector_source.hpp]: ah yes. I don't use c++ too often and forgot about that. fixed it now.

**Review by louwers — APPROVED** [review]
@xabbu42 Yup, that is what I meant.

https://en.cppreference.com/w/cpp/language/member_functions#Member_functions_with_cv-qualifiers

Thanks for your contribution!

**Pull request merged**
